src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +71,95 @@ def install_given_reqs(
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
-
-    return installed
+        # first try to install in parallel
+        installed_pool = _safe_pool_map(
+            partial(_single_install, install_args, in_subprocess=True),
+            list(to_install.values()))
+        if installed_pool:
+            installed = collections.OrderedDict(
+                zip(list(to_install.keys()), installed_pool))
+
+        for name, requirement in to_install.items():
+            if installed[name] is None:
+                installed_req = _single_install(
+                    install_args, requirement, in_subprocess=False)
+                installed[name] = installed_req  # type: ignore
+            elif isinstance(installed[name], BaseException):
+                raise installed[name]   # type: ignore
+
+    return [
+        i for i in installed.values()  # type: ignore
+        if isinstance(i, InstallationResult)
+    ]
+
+
+def _safe_pool_map(
+        func,               # type: Callable[[Any], Any]
+        iterable,           # type: Iterable[Any]
+):
+    # type: (...) -> Optional[List[Any]]
+    """
+    Safe call to Pool map, if Pool is not available return None
+    """
+    if not iterable:
+        return None
+
+    return map_multiprocess_ordered(func, iterable)
+
+
+def _single_install(
+        install_args,           # type: List[Any]
+        requirement,            # type: InstallRequirement
+        in_subprocess=False,    # type: bool
+):
+    # type: (...) -> Union[None, InstallationResult, BaseException]
+    """
+    Install a single requirement, returns InstallationResult
+    (to be called per requirement, either in parallel or serially)
+    """
+
+    # if we are running inside a subprocess,
+    # then only clean wheel installation is supported
+    if (in_subprocess and
+            (requirement.should_reinstall or not requirement.is_wheel)):
+        return None
+
+    if requirement.should_reinstall:
+        logger.info('Attempting uninstall: %s', requirement.name)
+        with indent_log():
+            uninstalled_pathset = requirement.uninstall(
+                auto_confirm=True
+            )
+    else:
+        uninstalled_pathset = None
+
+    try:
+        requirement.install(
+            install_args[0],   # install_options,
+            install_args[1],   # global_options,
+            **install_args[2]  # **kwargs
+        )
+    except (KeyboardInterrupt, SystemExit):
+        # always raise, we catch it in external loop
+        raise
+    except BaseException as ex:
```

Comment:
Rollback was only performed on `Exception` subclasses in the previous implementation. Is the change intentional?

Reply:
Yes that is on purpose, since we are running inside a sub-process and can get termination/process BaseException, if you feel strongly about it we can change to Exception.

Reply:
I’m only wondering whether we should keep the old semantic and not clean up on `BaseException` (like you also don’t in this implementation on `KeyboardInterrupt` and `SystemExit`.

Reply:
We should not cleanup on `KeyboardInterrupt` and `SystemExit` , otherwise we might fail to stop the process when requested. Regrading switching back to `except Exception`, I can't remember specific case where it is needed so I think the safest bet is to revert to `Exception` and add a comment. wdyt ?

Reply:
Yeah fine by me.